File: backend/app/schemas/catalog.py
from pydantic import BaseModel, Field


class CharacterSummary(BaseModel):
    id: str
    name: str
    rarity: int
    path: str
    element: str
    image_url: str | None = None


class NamedDescription(BaseModel):
    name: str
    description: str
    type: str | None = None
    image_url: str | None = None


class RelatedItem(BaseModel):
    id: str
    name: str
    rarity: str
    type: str
    description: str = ""
    image_url: str | None = None


class EntityRecommendation(BaseModel):
    id: str
    name: str
    kind: str
    image_url: str | None = None
    image_status: str = "official"
    set_type: str | None = None
    piece_count: int | None = None


class KnowledgeEntitySummary(BaseModel):
    id: str
    name: str
    kind: str
    subtitle: str = ""
    image_url: str | None = None
    image_status: str = "official"
    rarity: int | None = None
    path: str | None = None
    set_type: str | None = None
    piece_count: int | None = None


class RelatedEntity(KnowledgeEntitySummary):
    pass


class KnowledgeEntityDetail(KnowledgeEntitySummary):
    description: str = ""
    sections: list[NamedDescription] = Field(default_factory=list)
    related_entities: list[RelatedEntity] = Field(default_factory=list)
    source_url: str | None = None
    data_version: str | None = None


class LightconeLevelStats(BaseModel):
    level: int
    hp: int
    attack: int
    defence: int


class LightconeMaterial(RelatedItem):
    quantity: int


class LightconeDetail(KnowledgeEntityDetail):
    lore_description: str = ""
    effect_name: str = ""
    superimposition_effects: list[str] = Field(default_factory=list)
    level_stats: list[LightconeLevelStats] = Field(default_factory=list)
    promotion_materials: dict[int, list[LightconeMaterial]] = Field(default_factory=dict)


class RelicDetail(KnowledgeEntityDetail):
    set_type: str
    piece_count: int
    acquisition: str = ""
    set_effects: list[NamedDescription] = Field(default_factory=list)
    pieces: list[NamedDescription] = Field(default_factory=list)
    community_source_url: str | None = None


class CharacterDetail(CharacterSummary):
    description: str = ""
    energy: str | None = None
    stats: dict[str, str] = Field(default_factory=dict)
    skills: list[NamedDescription] = Field(default_factory=list)
    eidolons: list[NamedDescription] = Field(default_factory=list)
    related_items: list[RelatedItem] = Field(default_factory=list)
    recommended_lightcones: list[EntityRecommendation] = Field(default_factory=list)
    recommended_relics: list[EntityRecommendation] = Field(default_factory=list)
    story: str = ""
    community_source_url: str | None = None
    portrait_url: str | None = None
    element_icon_url: str | None = None
    path_icon_url: str | None = None
    source_url: str
    data_version: str | None = None


class ItemDetail(RelatedItem):
    background: str = ""
    sources: list[str] = Field(default_factory=list)
    pile_limit: int | None = None
    source_url: str | None = None
    related_entities: list[RelatedEntity] = Field(default_factory=list)


class CharacterListResponse(BaseModel):
    items: list[CharacterSummary]
    total: int


class ItemListResponse(BaseModel):
    items: list[RelatedItem]
    total: int


class KnowledgeEntityListResponse(BaseModel):
    items: list[KnowledgeEntitySummary]
    total: int

File: frontend/src/types/catalog.ts
export interface CharacterSummary {
  id: string
  name: string
  rarity: number
  path: string
  element: string
  image_url?: string
}

export interface NamedDescription {
  name: string
  description: string
  type?: string
  image_url?: string
}

export interface RelatedItem {
  id: string
  name: string
  rarity: string
  type: string
  description: string
  image_url?: string
}

export interface EntityRecommendation {
  id: string
  name: string
  kind: string
  image_url?: string
  image_status: 'official' | 'category_placeholder'
  set_type?: 'cavern' | 'planar'
  piece_count?: number
}

export interface KnowledgeEntitySummary {
  id: string
  name: string
  kind: 'characters' | 'lightcones' | 'relics' | 'items' | 'monsters'
  subtitle: string
  image_url?: string
  image_status: 'official' | 'category_placeholder'
  rarity?: number
  path?: string
  set_type?: 'cavern' | 'planar'
  piece_count?: number
}

export interface KnowledgeEntityDetail extends KnowledgeEntitySummary {
  description: string
  sections: NamedDescription[]
  related_entities: KnowledgeEntitySummary[]
  source_url?: string
  data_version?: string
}

export interface CharacterDetail extends CharacterSummary {
  description: string
  energy?: string
  stats: Record<string, string>
  skills: NamedDescription[]
  eidolons: NamedDescription[]
  related_items: RelatedItem[]
  recommended_lightcones: EntityRecommendation[]
  recommended_relics: EntityRecommendation[]
  story: string
  community_source_url?: string
  portrait_url?: string
  element_icon_url?: string
  path_icon_url?: string
  source_url: string
  data_version?: string
}

export interface ItemDetail extends RelatedItem {
  background: string
  sources: string[]
  pile_limit?: number
  source_url?: string
  related_entities: KnowledgeEntitySummary[]
}

export interface LightconeLevelStats {
  level: number
  hp: number
  attack: number
  defence: number
}

export interface LightconeMaterial extends RelatedItem {
  quantity: number
}

export interface LightconeDetail extends KnowledgeEntityDetail {
  lore_description: string
  effect_name: string
  superimposition_effects: string[]
  level_stats: LightconeLevelStats[]
  promotion_materials: Record<string, LightconeMaterial[]>
}

export interface RelicDetail extends KnowledgeEntityDetail {
  set_type: 'cavern' | 'planar'
  piece_count: number
  acquisition: string
  set_effects: NamedDescription[]
  pieces: NamedDescription[]
  community_source_url?: string
}
